packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -143,203 +192,138 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
-        uint256 _vaultId,
-        uint256 _wPowerPerpAmountToMint,
-        uint256 _collateralAmount,
-        uint256 _wPowerPerpAmountToSell,
-        uint256 _minToReceive
-    ) external payable {
-        IWPowerPerp(wPowerPerp).transferFrom(msg.sender, address(this), _wPowerPerpAmountToSell);
+    function flashswapSellLongWMint(FlashSellLongWMintParams calldata _params) external payable {
+        IWPowerPerp(wPowerPerp).transferFrom(msg.sender, address(this), _params.wPowerPerpAmountToSell);
 
-        // flahswap and mint short position
+        // flashswap and mint short position
         _exactInFlashSwap(
             wPowerPerp,
             weth,
             IUniswapV3Pool(wPowerPerpPool).fee(),
-            _wPowerPerpAmountToMint.add(_wPowerPerpAmountToSell),
-            _minToReceive,
-            uint8(FLASH_SOURCE.FLASH_SELL_LONG_W_MINT),
-            abi.encodePacked(_vaultId, _wPowerPerpAmountToMint, _collateralAmount)
+            _params.wPowerPerpAmountToMint.add(_params.wPowerPerpAmountToSell),
+            _params.minToReceive,
+            uint8(CALLBACK_SOURCE.FLASH_SELL_LONG_W_MINT),
+            abi.encode(_params)
         );
+
+        payable(msg.sender).sendValue(address(this).balance);
+    }
+
+    /**
+     * @notice close short position with user Uniswap v3 LP NFT
+     * @dev user should approve this contract for Uni NFT transfer
+     */
+    function closeShortWithUserNft(CloseShortWithUserNftParams calldata _params) external {
+        INonfungiblePositionManager(nonfungiblePositionManager).safeTransferFrom(
+            msg.sender,
+            address(this),
+            _params.tokenId
+        );
+
+        _closeUniLp(
+            closeUniLpParams({
+                vaultId: _params.vaultId,
+                tokenId: _params.tokenId,
+                liquidityPercentage: _params.liquidityPercentage,
+                wPowerPerpAmountToBurn: _params.wPowerPerpAmountToBurn,
+                collateralToWithdraw: _params.collateralToWithdraw,
+                limitPriceEthPerPowerPerp: _params.limitPriceEthPerPowerPerp,
+                amount0Min: _params.amount0Min,
+                amount1Min: _params.amount1Min
+            })
+        );
+
+        IWETH9(weth).withdraw(IWETH9(weth).balanceOf(address(this)));
+        payable(msg.sender).sendValue(address(this).balance);
+    }
+
+    function flashloanCloseVaultLpNft(FlashloanCloseVaultLpNftParam calldata _params) external payable {
+        _flashLoan(
+            weth,
+            _params.collateralToFlashloan,
+            uint8(CALLBACK_SOURCE.FLASHLOAN_CLOSE_VAULT_LP_NFT),
+            abi.encode(
+                closeUniLpParams({
+                    vaultId: _params.vaultId,
+                    tokenId: _params.tokenId,
+                    liquidityPercentage: uint256(1e18),
+                    wPowerPerpAmountToBurn: _params.wPowerPerpAmountToBurn,
+                    collateralToWithdraw: _params.collateralToFlashloan,
```

Comment:
this should be a param right?  (can withdraw partial like closeShortWithUserNft()